### PR TITLE
Add zoom smoothing

### PR DESCRIPTION
**Pull request by 341101**
See https://github.com/Plonq/bevy_panorbit_camera/pull/29 for previous discussion

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -464,24 +484,18 @@ fn pan_orbit_camera(
...
+            let line_delta = -scroll_line * pan_orbit.target_radius * 0.2;
+            let pixel_delta = -scroll_pixel * pan_orbit.target_radius * 0.2;
+            // Add the calculated deltas to the target radius
+            pan_orbit.target_radius = util::apply_zoom_limits(
+                pan_orbit.target_radius + line_delta + pixel_delta,
+                pan_orbit.zoom_upper_limit,
+                pan_orbit.zoom_lower_limit,
+            );
+            // If it is pixel-based scrolling, it is added directly to the target
+            pan_orbit.radius = pan_orbit.radius.map(|radius| radius + pixel_delta);
```
Comment (Plonq) [inline src/lib.rs:498]: Going to comment on the new PR now.

You aren't applying zoom limits to this line - pixel scroll delta is being applied directly to radius without limits.

File: src/lib.rs [hunk 2]
```diff
@@ -517,6 +531,25 @@ fn pan_orbit_camera(
...
+                has_moved = true;
+                // Smoothing for line-based scrolling
+                let t = 1.0 - pan_orbit.zoom_smoothness;
+                let mut next_radius = radius.lerp(&pan_orbit.target_radius, &t);
+                if approx_equal(next_radius, pan_orbit.target_radius) {
+                    next_radius = pan_orbit.target_radius
+                }
+                if let Projection::Orthographic(ref mut p) = *projection {
+                    p.scale = next_radius;
+                }
```
Comment (Plonq) [inline src/lib.rs:546]: As stated in other PR, you need to do this for pixel scrolling too

**Review by Plonq — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 2 above)
Comment (341101) [reply src/lib.rs:546]: I added the values of both line scrolling and pixel scrolling to the target_radius.

Comment (341101) [reply src/lib.rs:551]: > As stated in other PR, you need to do this for pixel scrolling too

I added the values of both line scrolling and pixel scrolling to the target_radius.

File: src/lib.rs (hunk 1 above)
Comment (341101) [reply src/lib.rs:498]: > Going to comment on the new PR now.
> 
> You aren't applying zoom limits to this line - pixel scroll delta is being applied directly to radius without limits.

Perhaps we could add another function, like this one?
```
pub(super) fn interpolate_and_check_approx_f32(target: f32, current: f32, smoothness: f32) -> f32 {
    let t = 1.0 - smoothness;
    let mut new_value = current.lerp(&target, &t);
    if approx_equal(new_value, target) {
        new_value = target;
    }
    new_value
}
```

**Comment by 341101** [comment]
I have bound the scale of the Orthographic camera to radius, which means the behavior of the Orthographic camera in the PR code will be different. Should the original behavior be preserved?

[origin](https://github.com/Plonq/bevy_panorbit_camera/assets/35809384/5ccd0146-a9dd-4f55-ac80-fdfd60a0ef1f)

[new](https://github.com/Plonq/bevy_panorbit_camera/assets/35809384/6a99ecb8-b73b-4c37-9076-9c0be2a28897)

**Comment by Plonq** [comment]
> I have bound the scale of the Orthographic camera to radius, which means the behavior of the Orthographic camera in the PR code will be different. Should the original behavior be preserved?

Yes original behaviour should be preserved, to avoid clipping like your video shows. In other words, for orthographic cameras, the radius should never change when zooming.

**Comment by 341101** [comment]
> It's almost there - both pixel and line scrolling work, including for ortho cameras. Let's just make sure radius is not modified for orthographic cameras, and then I think it'll be ready.

I have added the `scale` and `target_scale` fields, restoring the behavior of the orthographic camera. Additionally, I have modified the code related to Constraints, reusing the `apply_limits` function (previously `apply_zoom_limits`).

**Comment by Plonq** [comment]
> > It's almost there - both pixel and line scrolling work, including for ortho cameras. Let's just make sure radius is not modified for orthographic cameras, and then I think it'll be ready.
> 
> I have added the `scale` and `target_scale` fields, restoring the behavior of the orthographic camera. Additionally, I have modified the code related to Constraints, reusing the `apply_limits` function (previously `apply_zoom_limits`).

This is getting a bit out of hand, and I think there's too much scope creep now. I don't think adding new fields for scale is a good idea, when radius can be used for both. But it's not obvious how to structure things accordingly.

How about this. Let's revert back to before you differentiated between pixel and line scrolling. I.e. apply smoothing across the board. Once merged, I'll look into solving the lag issue with smoothing the pixel scrolling separately.

Don't worry, your work won't be wasted!

**Comment by 341101** [comment]
> > > It's almost there - both pixel and line scrolling work, including for ortho cameras. Let's just make sure radius is not modified for orthographic cameras, and then I think it'll be ready.
> > 
> > 
> > I have added the `scale` and `target_scale` fields, restoring the behavior of the orthographic camera. Additionally, I have modified the code related to Constraints, reusing the `apply_limits` function (previously `apply_zoom_limits`).
> 
> This is getting a bit out of hand, and I think there's too much scope creep now. I don't think adding new fields for scale is a good idea, when radius can be used for both. But it's not obvious how to structure things accordingly.
> 
> How about this. Let's revert back to before you differentiated between pixel and line scrolling. I.e. apply smoothing across the board. Once merged, I'll look into solving the lag issue with smoothing the pixel scrolling separately.
> 
> Don't worry, your work won't be wasted!

I think in perspective and orthographic cameras, radius and scale have different roles. Merging them could lead to confusion. For clarity and ease of use.

**Comment by Plonq** [comment]
> I think in perspective and orthographic cameras, radius and scale have different roles. Merging them could lead to confusion. Even in orthographic cameras, we still need radius to represent the position of the camera. For clarity and ease of use, it's better to keep them separate.

After thinking about it some more, and investigating what it would take to use radius for both, I have come to agree overall. There's still one thing I don't like about splitting the values, and that is if you want to control the 'zoom' of the camera, and dynamically switch between normal and orthographic projection, you would also need to switch between radius and scale. But that's likely an edge case, and isn't all that bad.

I'm happy with the PR as it is. Let's get this merged!

**Review by Plonq — APPROVED** [review]

**Comment by 341101** [comment]
> > I think in perspective and orthographic cameras, radius and scale have different roles. Merging them could lead to confusion. Even in orthographic cameras, we still need radius to represent the position of the camera. For clarity and ease of use, it's better to keep them separate.
> 
> After thinking about it some more, and investigating what it would take to use radius for both, I have come to agree overall. There's still one thing I don't like about splitting the values, and that is if you want to control the 'zoom' of the camera, and dynamically switch between normal and orthographic projection, you would also need to switch between radius and scale. But that's likely an edge case, and isn't all that bad.
> 
> I'm happy with the PR as it is. Let's get this merged!
> 
> edit: just need a rustfmt. Also, it would be good to add some tests for the new util functions.

I've run rustfmt and have added tests for `apply_limits` and `approx_equal`.

**Comment by 341101** [comment]
> > I think in perspective and orthographic cameras, radius and scale have different roles. Merging them could lead to confusion. Even in orthographic cameras, we still need radius to represent the position of the camera. For clarity and ease of use, it's better to keep them separate.
> 
> After thinking about it some more, and investigating what it would take to use radius for both, I have come to agree overall. There's still one thing I don't like about splitting the values, and that is if you want to control the 'zoom' of the camera, and dynamically switch between normal and orthographic projection, you would also need to switch between radius and scale. But that's likely an edge case, and isn't all that bad.
> 
> I'm happy with the PR as it is. Let's get this merged!
> 
> edit: just need a rustfmt. Also, it would be good to add some tests for the new util functions.

I've run rustfmt and have added tests for `apply_limits` and `approx_equal`.

**Comment by Plonq** [comment]
Thanks again for the contribution! I'll get a release out soon.
